integration-tests/jpa-postgresql-withxml/src/main/java/io/quarkus/it/jpa/postgresql/JPAFunctionalityTestEndpoint.java

```diff
@@ -0,0 +1,172 @@
+package io.quarkus.it.jpa.postgresql;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.UUID;
+
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.EntityTransaction;
+import javax.persistence.TypedQuery;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Root;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.postgresql.jdbc.PgSQLXML;
+
+/**
+ * Various tests covering JPA functionality. All tests should work in both standard JVM and in native mode.
+ */
+@WebServlet(name = "JPATestBootstrapEndpoint", urlPatterns = "/jpa-withxml/testfunctionality")
+public class JPAFunctionalityTestEndpoint extends HttpServlet {
+
+    @Inject
+    EntityManagerFactory entityManagerFactory;
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        try {
+            doStuffWithHibernate(entityManagerFactory);
+        } catch (Exception e) {
+            reportException("An error occurred while performing Hibernate operations", e, resp);
+        }
+        resp.getWriter().write("OK");
+    }
+
+    /**
+     * Lists the various operations we want to test for:
+     */
+    private static void doStuffWithHibernate(EntityManagerFactory entityManagerFactory) {
+
+        //Cleanup any existing data:
+        deleteAllPerson(entityManagerFactory);
+
+        //Store some well known Person instances we can then test on:
+        storeTestPersons(entityManagerFactory);
+
+        //Load all persons and run some checks on the query results:
+        verifyListOfExistingPersons(entityManagerFactory);
+
+        //Try a JPA named query:
+        verifyJPANamedQuery(entityManagerFactory);
+
+        deleteAllPerson(entityManagerFactory);
+
+        useXMLParsing(entityManagerFactory);
+
+    }
+
+    private static void useXMLParsing(EntityManagerFactory entityManagerFactory) {
+        //This isn't very useful but triggers the rules in io.quarkus.jdbc.postgresql.runtime.graal.SQLXLMFeature
+        PgSQLXML xmlType = new PgSQLXML(null);
+        try {
+            xmlType.setResult(null);
+        } catch (SQLException throwables) {
+            throwables.printStackTrace();
+        }
```

Comment:
From what I understand, you are not actually testing that XML parsing works? Just that when we call this method (and it fails) some XML classes are included. There could still be side-effects of the substitutions that render the feature unusable, especially since we replaced large parts of the `PgSQLXML` class.

Could this be replaced by an actual test that makes use of the XML parsing? For example by running a native query that uses the `PgSQLXML` type in some way, or by declaring a UserType that uses `PgSQLXML` somehow?

Otherwise, it feels like all the efforts you did to make sure people can still use this feature could be for nothing, and that would be a shame...

FWIW, I would also get rid of all the unrelated complexity in the entities involved in this endpoint, but that's not strictly necessary, I suppose.

Reply:
Agreed, I'll try. I must admit I wanted to but it got late and it all seemed too good to postpone further :)

Reply:
This is done, and indeed it wasn't all working yet.